enhancements/update/synchronized-upgrades.md

```diff
@@ -0,0 +1,223 @@
+### API Extensions
+
+TBD
+
+### Implementation Details/Notes/Constraints [optional]
+
+TBD
+
+### Risks and Mitigations
+
+TBD
+
+## Design Details
+
+### Open Questions
+
+Basically everything...
+
+The general idea is:
+
+- We can set some things up at install time (eg, creating credentials
+  to allow certain operators in the two clusters to talk to each
+  other).
+
+- As part of the DPU security model, the tenant cluster cannot have
+  any power over the infra cluster. (In particular, it can't be
+  possible for an administrator in the tenant cluster to force the
+  infra cluster to upgrade/downgrade to any particular version.) Thus,
+  the upgrade must be initiated on the infra cluster side, and the
+  infra side will tell the tenant cluster to upgrade as well.
+  (Alternatively, the upgrade must be simultaneous-ish-ly initiated in
+  both clusters, if we don't want the infra cluster to have to have a
+  credential that lets it initiate an upgrade in the tenant cluster.)
+
+- An upgrade should not be able to start unless both clusters are able
+  to upgrade.
+
+  - In particular:
+
+    - There can be no `Upgradeable: False` operators in either cluster.
+
+    - The version to upgrade to must be available to both clusters
+      (ie, it must be available for both x86 and ARM).
+
+  - This could be implemented via some sort of "dpu-cluster-upgrade"
+    operator running in both clusters, where the two operators
+    communicate with each other and set their `Upgradeable` state to
+    reflect the state of the other cluster. If the
+    "dpu-cluster-upgrade" operator was placed before every other
+    operator in upgrade priority, then it could also block disallowed
+    upgrades, by failing its own upgrade, eg if an admin tries to
+    upgrade one cluster without the other, or tries to upgrade the two
+    clusters to different versions.
+
+    - (Or should it be possible to do z-stream upgrades of the tenant
+      cluster without bothering to upgrade the infra cluster too?)
+
+- The two clusters upgrade all of the operators up to MCO in parallel.
+
+- Whichever cluster reaches the MCO upgrade first needs to wait for
+  the other cluster to get there before proceeding. The two MCOs then
+  need to coordinate to complete the upgrade: First, they have to
+  agree on what order the physical hosts will be upgraded in. Second,
+  for each physical host, they have to properly synchronize the
+  upgrades of its infra node and its tenant node.
+
+  - More specifically, for each physical host, in some order:
+
+    - The Infra MCO will cordon and drain that host's infra node, and
+      the Tenant MCO will cordon and drain that host's tenant node.
+      (This can happen in parallel.)
+
+    - The Infra MCO will then upgrade the infra node (causing it to
+      reboot and temporarily break network connectivity to the tenant
+      node).
+
+    - Once the infra node upgrade completes, the Tenant MCO will
+      reboot and upgrade the tenant node.
```

Comment:
Does rebooting the tenant node implicitly restart the infra node or does the infra node remain online during a tenant node reboot? Does this vary by either infra node or tenant node hardware or even configuration such as BIOS settings?

Reply:
In our lab, rebooting the tenant does not restart the NIC, and AFAIK this does not vary by hardware. (ie, ACPI/EFI say that on reboot you send a PCI reset but you don't cut power to the bus.) IIRC an earlier (prerelease) version of the BF-2 did reboot on reboot (I guess because it was designed to reboot on PCI reset) but the GA ones do not.

If there are situations where the NIC reboots on host reboot (including the possibility of other non-BF-2 DPUs in the future that behave that way), then it would only be a small change to the procedure here; instead of draining both nodes, rebooting the NIC, and then rebooting the host, we would drain both nodes, then reboot the host without explicitly rebooting the NIC.